transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,290 @@ fn test_input_rules_verify_max_tombstone() {
+            let amount_shared_secret =
+                MaskedAmount::compute_amount_shared_secret(block_version, &tx_out_shared_secret)
+                    .unwrap();
+            let (amount, _) = txo
+                .masked_amount
+                .as_ref()
+                .unwrap()
+                .get_value_from_amount_shared_secret(&amount_shared_secret)
+                .unwrap();
+            assert_eq!(amount.value, 1000);
+            assert_eq!(*amount.token_id, 0);
+            RevealedTxOut {
+                tx_out: txo.clone(),
+                amount_shared_secret: amount_shared_secret.to_vec(),
+            }
+        })
+        .collect();
+
+    // Add a fractional output, by doubling one of the revealed tx outs.
+    // So this means we are filling the 1/2 of the request.
+    // Before we add a fractional change output, this is ill-formed, we can only
+    // verify fractional output rules if there is a fractional change output,
+    // because that is how we deduce the fill fraction.
+    get_first_rules_mut(&mut tx)
+        .fractional_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(2000, 0.into()),
+        ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputsNotExpected)
+    );
+
+    // Add a fractional input, also by doubling one of the revealed tx outs.
+    // This means the fill fraction is 1/2, which we are satisfying.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(2000, 0.into()),
+    ));
+
+    // Check that the Tx is following input rules
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Lets try imposing a limit on the change value
+    // This is larger than 1000 so we should still be valid.
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 1010;
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Lets try imposing a smaller limit that should cause things to fail
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 990;
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::RealChangeOutputAmountExceedsLimit)
+    );
+
+    // Remove limit altogether, we should be valid again
+    get_first_rules_mut(&mut tx).max_allowed_change_value = 0;
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Change the frational change output to be 3000. This means the implied fill
+    // fraction is now 2/3, since the real change output only returns
+    // 1/3 of this. The tx should then be invalid because we are only filling 1/2 of
+    // the fractional output that was required and not 2/3 of it.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(3000, 0.into()),
```

Comment:
:thought_balloon: By the time I got down to here it's hard to recall what the overall state of the TX is.  I think this would be better to have a common test set up method and then individual test cases that focused on what they want the state to be over this recipe/workflow style test

Reply:
I agree that having better test infrastructure would make the tests better -- most of the test infrastructure in transaction-core is trash.

I've also gotten a lot of feedback that these PRs are too big. I don't think we can justify introducing new test infrastructure in this PR, but perhaps in a later PR.